tools/nsc/main.cpp

```diff
@@ -0,0 +1,155 @@
+			return false;
+		}
+		std::string output_filepath = "";
+		for (auto i = 1; i < m_arguments.size(); i++)
+		{
+			if (m_arguments[i] == "-no-nbl-builtins")
+			{
+				// unmount builtins
+				m_logger->log("Unmounting builtins.");
+				m_system->unmountBuiltins();
+				no_nbl_builtins = true;
+				m_arguments.erase(m_arguments.begin() + i - 1);
+				i--;
+			}
+			else if (m_arguments[i] == "-Fo")
+			{
+				// get filepath to save output to
+				if (i + 1 < argc)
+				{
+					output_filepath = argv[i + 1];
+					m_logger->log("Compiled shader code will be saved to " + output_filepath);
+					m_arguments.erase(m_arguments.begin() + i, m_arguments.begin() + i + 1);
+					i--;
+				}
+				else
+				{
+					m_logger->log("Incorrect arguments. Expecting filename after -Fo.", ILogger::ELL_ERROR);
+				}
+			}
+
+		}
+
+#ifndef NBL_EMBED_BUILTIN_RESOURCES
+		if (!no_nbl_builtins) {
+			m_system->unmountBuiltins();
+			no_nbl_builtins = true;
+			m_logger->log("nsc.exe was compiled with builtin resources disabled. Force enabling -no-nbl-builtins.", ILogger::ELL_WARNING);
+		}
+#endif
+		string shader_code = open_shader_file(file_to_compile);
+		auto compilation_result = compile_shader(shader_code, file_to_compile);
+
+		// writie compiled shader to file as bytes
+		if (compilation_result && !output_filepath.empty()) {
+			std::fstream output_file(output_filepath, std::ios::out | std::ios::binary);
+			output_file.write((const char*)compilation_result->getContent()->getPointer(), compilation_result->getContent()->getSize());
+			output_file.close();
+			m_logger->log("Shader compilation successful.");
+		}
+		else {
+			m_logger->log("Shader compilation failed.", ILogger::ELL_ERROR);
+		}
+		return true;
+	}
+
+	void workLoopBody() override {}
+
+	bool keepRunning() override { return false; }
+
+
+private:
+
+	core::smart_refctd_ptr<ICPUShader> compile_shader(std::string& shader_code, std::string_view sourceIdentifier) {
+		constexpr uint32_t WorkgroupSize = 256;
+		constexpr uint32_t WorkgroupCount = 2048;
+		const string WorkgroupSizeAsStr = std::to_string(WorkgroupSize);
+		const IShaderCompiler::SPreprocessorOptions::SMacroDefinition WorkgroupSizeDefine = { "WORKGROUP_SIZE",WorkgroupSizeAsStr };
+
+		smart_refctd_ptr<CHLSLCompiler> hlslcompiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(m_system));
+
+		CHLSLCompiler::SOptions options = {};
+		// want as much debug as possible
+		options.stage = asset::IShader::E_SHADER_STAGE::ESS_UNKNOWN;
+		options.debugInfoFlags = IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_LINE_BIT;
+		// this lets you source-level debug/step shaders in renderdoc
+		//if (physDev->getLimits().shaderNonSemanticInfo)
+		//options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_NON_SEMANTIC_BIT;
+		// if you don't set the logger and source identifier you'll have no meaningful errors
```

Comment:
dont set those

Reply:
I remember that the code that emits `#pragma wave dxc_compile_flags` is conditional, it only adds the pragma to compiled code if this debug flag `IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_LINE_BIT` or `EDIF_NON_SEMANTIC_BIT` or `EDIF_SOURCE_BIT` is on 

How should we go about this? Do you want the pragma to always be emitted, or only when user gives the corresponding `-fspv-debug=file|source|line` arg in command line when running NSC tool?

Reply:
you can always emit it, its just that DXC will "loose" the original source and comments AFTER wave is done preprocessing without the `-fspv-debug` which is fine... don't make emission of `#pragma wave dxc_compile_flags` conditional, always do it